File: src/api/git.ts
import { invoke } from "@tauri-apps/api/core";

import type { GitBranch, GitCommit, GitStatus } from "./types";

export function getGitStatus(path: string): Promise<GitStatus> {
  return invoke<GitStatus>("get_git_status", { path });
}

export function stageAll(path: string): Promise<void> {
  return invoke<void>("stage_all", { path });
}

export function unstageAll(path: string): Promise<void> {
  return invoke<void>("unstage_all", { path });
}

export function fetch(path: string): Promise<void> {
  return invoke<void>("fetch", { path });
}

export function commit(path: string, message: string): Promise<void> {
  return invoke<void>("commit", { path, message });
}

export function pull(path: string): Promise<void> {
  return invoke<void>("pull", { path });
}

export function push(path: string): Promise<void> {
  return invoke<void>("push", { path });
}

export function getGitDiff(path: string, staged = false): Promise<string> {
  return invoke<string>("get_git_diff", { path, staged });
}

export function listBranches(path: string): Promise<GitBranch[]> {
  return invoke<GitBranch[]>("list_branches", { path });
}

export function switchBranch(path: string, branch: string): Promise<void> {
  return invoke<void>("switch_branch", { path, branch });
}

export function createBranch(path: string, branch: string): Promise<void> {
  return invoke<void>("create_branch", { path, branch });
}

export function deleteBranch(path: string, branch: string): Promise<void> {
  return invoke<void>("delete_branch", { path, branch });
}

export function getGitLog(path: string, limit = 30): Promise<GitCommit[]> {
  return invoke<GitCommit[]>("git_log", { path, limit });
}

File: src-tauri/src/commands/git.rs
use std::path::PathBuf;

use crate::git::{self, GitBranch, GitCommit, GitStatus};

#[tauri::command]
pub async fn get_git_status(path: PathBuf) -> Result<GitStatus, String> {
    tauri::async_runtime::spawn_blocking(move || git::status(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn stage_all(path: PathBuf) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::stage_all(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn unstage_all(path: PathBuf) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::unstage_all(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn fetch(path: PathBuf) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::fetch(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn commit(path: PathBuf, message: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::commit(&path, &message))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn pull(path: PathBuf) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::pull(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn push(path: PathBuf) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::push(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn get_git_diff(path: PathBuf, staged: bool) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || git::diff(&path, staged))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn list_branches(path: PathBuf) -> Result<Vec<GitBranch>, String> {
    tauri::async_runtime::spawn_blocking(move || git::branches(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn switch_branch(path: PathBuf, branch: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::switch_branch(&path, &branch))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn create_branch(path: PathBuf, branch: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::create_branch(&path, &branch))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn delete_branch(path: PathBuf, branch: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || git::delete_branch(&path, &branch))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn git_log(path: PathBuf, limit: u32) -> Result<Vec<GitCommit>, String> {
    tauri::async_runtime::spawn_blocking(move || git::log(&path, limit))
        .await
        .map_err(|error| error.to_string())?
}
